File: java/src/binary_heap/HeapDownTest.java
package binary_heap;

import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;

public class HeapDownTest {
	@Test
	public void testRandomCases() {
		for (int test = 0; test < 1_000; test++) {
			var td = new randomTestData();

			Assert.assertTrue(randomTestData.minHeapIsOK(td.heap));

			td.heap[td.randomIndex] += td.randomValue;
			var actual = td.heap.clone();
			HeapDown.heapDown(actual, actual.length, td.randomIndex);

			if (!randomTestData.minHeapIsOK(actual)) {
				Assert.fail(
					String.format("heapDown(%s, %s, %s): not a valid heap after the call:\n%s",
						Arrays.toString(td.heap), td.randomIndex, td.heap.length, Arrays.toString(actual)
					)
				);
			}

			if (!randomTestData.haveSameElements(actual, td.heap)) {
				Assert.fail(
					String.format("heapDown(%s, %s, %s): not same elements after the call:\n%s",
						Arrays.toString(td.heap), td.randomIndex, td.heap.length, Arrays.toString(actual)
					)
				);
			}
		}
	}
}

File: python/binary_heap/heap_down_test.py
from heap_down import heap_down
from bh_random_test_data import RandomTestData
from bh_random_test_data import min_heap_is_ok


def test_heap_down_random_cases():
	for _ in range(1_000):
		td = RandomTestData()

		assert min_heap_is_ok(td.heap)

		td.heap[td.random_index] += td.randomValue
		got = td.heap.copy()
		heap_down(got, len(got), td.random_index)

		assert min_heap_is_ok(got), \
			f'heap_down({td.heap}, {td.random_index}): not a valid heap after the call:\n{got}'

		assert sorted(got) == sorted(td.heap), \
			f'heap_down({td.heap}, {td.random_index}): not same elements after the call:\n{got}'
